module/VuFind/src/VuFind/Db/Service/UserCardService.php

```diff
@@ -0,0 +1,324 @@
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+
+namespace VuFind\Db\Service;
+
+use DateTime;
+use VuFind\Auth\ILSAuthenticator;
+use VuFind\Config\AccountCapabilities;
+use VuFind\Db\Entity\UserCardEntityInterface;
+use VuFind\Db\Entity\UserEntityInterface;
+use VuFind\Db\Table\DbTableAwareInterface;
+use VuFind\Db\Table\DbTableAwareTrait;
+
+use function count;
+use function is_int;
+
+/**
+ * Database service for UserCard.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+class UserCardService extends AbstractDbService implements
+    DbServiceAwareInterface,
+    DbTableAwareInterface,
+    UserCardServiceInterface
+{
+    use DbServiceAwareTrait;
+    use DbTableAwareTrait;
+
+    /**
+     * Constructor
+     *
+     * @param ILSAuthenticator    $ilsAuthenticator ILS authenticator
+     * @param AccountCapabilities $capabilities     Account capabilities configuration
+     */
+    public function __construct(
+        protected ILSAuthenticator $ilsAuthenticator,
+        protected AccountCapabilities $capabilities
+    ) {
+    }
+
+    /**
+     * Get user_card rows with insecure catalog passwords
+     *
+     * @return UserCardEntityInterface[]
+     */
+    public function getInsecureRows(): array
+    {
+        return iterator_to_array($this->getDbTable('UserCard')->getInsecureRows());
+    }
+
+    /**
+     * Get all library cards associated with the user.
+     *
+     * @param int|UserEntityInterface $user        User object or identifier
+     * @param ?int                    $id          Optional card ID filter
+     * @param ?string                 $catUsername Optional catalog username filter
+     *
+     * @return UserCardEntityInterface[]
+     */
+    public function getLibraryCards(int|UserEntityInterface $user, ?int $id = null, ?string $catUsername = null): array
+    {
+        if (!$this->capabilities->libraryCardsEnabled()) {
+            return [];
+        }
+        $userCard = $this->getDbTable('UserCard');
+        $criteria = [
+            'user_id' => is_int($user) ? $user : $user->getId(),
+        ];
+        if ($id) {
+            $criteria['id'] = $id;
+        }
+        if ($catUsername) {
+            $criteria['cat_username'] = $catUsername;
+        }
+        return iterator_to_array($userCard->select($criteria));
+    }
+
+    /**
+     * Get or create library card data.
+     *
+     * @param int|UserEntityInterface $user User object or identifier
+     * @param ?int                    $id   Card ID to fetch (or null to create a new card)
+     *
+     * @return UserCardEntityInterface Card data if found, null otherwise
```

Comment:
`?UserCardEntityInterface`

Reply:
Updated to match the interface, as noted above.